src/WebEid.Security/Nonce/NonceGeneratorBuilder.cs

```diff
@@ -0,0 +1,57 @@
+namespace WebEID.Security.Nonce
+{
+    using System;
+    using System.Security.Cryptography;
+    using Cache;
+    using Util;
+
+    /// <summary>
+    /// Builder for constructing NonceGenerator instances.
+    /// </summary>
+    public class NonceGeneratorBuilder
+    {
+        private readonly RandomNumberGenerator randomNumberGenerator;
+        private readonly ICache<DateTime> cache;
+        private TimeSpan ttl = TimeSpan.FromMinutes(5);
+
+        /// <summary>
+        /// Initializes a new instance of NonceGeneratorBuilder
+        /// </summary>
+        /// <param name="randomNumberGenerator">The source of random bytes for the nonce.</param>
+        /// <param name="cache">The cache where generated nonce values will be stored. The key of the cache is the Base64-encoded nonce and the value is the nonce expiry time.</param>
+        public NonceGeneratorBuilder(RandomNumberGenerator randomNumberGenerator, ICache<DateTime> cache)
+        {
+            this.randomNumberGenerator = randomNumberGenerator ?? throw new ArgumentNullException(nameof(cache), "Secure random generator must not be null");
+            this.cache = cache ?? throw new ArgumentNullException(nameof(cache), "Cache must not be null");
+        }
+
+        /// <summary>
+        /// Override default nonce time-to-live duration.
+        /// When the time-to-live passes, the nonce is considered to be expired.
+        /// </summary>
+        /// <param name="duration">time-to-live duration</param>
+        /// <returns>current builder instance</returns>
+        public NonceGeneratorBuilder WithNonceTtl(TimeSpan duration)
+        {
+            this.ttl = duration;
+            return this;
+        }
+        
+        /// <summary>
+        /// Builds the new nonce generator instance.
+        /// </summary>
+        /// <returns>new nonce generator instance</returns>
+        public INonceGenerator Build()
+        {
+            return new NonceGenerator(this.cache, this.randomNumberGenerator, this.ttl);
+        }
+
+        public static void RequirePositiveDuration(TimeSpan duration, string fieldName)
+        {
+            if (duration.IsNegative() || duration.IsZero())
```

Comment:
I would merge `IsNegative()` and `IsZero()` and use `<= 0` in `IsNegativeOrZero()`:

```suggestion
            if (duration.IsNegativeOrZero())
```